### PR TITLE
fix: move licensing checks to worker process

### PR DESCRIPTION
**Pull request by kopancek**
Moving licensing checks to a singleton worker job. This way we make sure that only one license check API call is being made on each 12-hour tick.

Previously this lived in a frontend service. If multiple frontend services were running, this would fire multiple requests, which is undesirable from many points of view - waste of resources, waste of bandwidth, potential concurrency issues. 

Having it fire multiple times would also break an assumption that we have in the license check handler on dotcom. For checking if same site_id is used on multiple instances, we assume that the time difference between license check calls would be roughly 10-12 hours. Calling twice in the same second breaks this assumption.

## Test plan

Tested locally that the worker starts

### CODE REVIEW

File: enterprise/cmd/frontend/internal/licensing/init/init.go [hunk 1]
```diff
@@ -157,38 +157,14 @@ func Init(
...
-			db: db,
-		})
-	})
 	if envvar.SourcegraphDotComMode() {
 		goroutine.Go(func() {
 			productsubscription.StartCheckForUpcomingLicenseExpirations(logger, db)
 		})
 		goroutine.Go(func() {
 			productsubscription.StartCheckForAnomalousLicenseUsage(logger, db)
 		})
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/licensing/init/init.go:166]: The `producsubscription` package is enterprise/internal, did not want to get deep into that can of worms.

**Comment by Strum355** [comment]
> Since worker is a singleton

Worker isnt a singleton, its designed to be able to have multiple running if you want to scale out other jobs. Cant remember how jobs are chosen though, @efritz do you remember how its configured and what the default is e.g. run all jobs by default

**Comment by efritz** [comment]
See `WORKER_JOB_ALLOWLIST` and `WORKER_JOB_BLOCKLIST`

**Comment by kopancek** [comment]
> See `WORKER_JOB_ALLOWLIST` and `WORKER_JOB_BLOCKLIST`

So based on the [docs](https://docs.sourcegraph.com/admin/workers#3-split-jobs-and-scale-independently) depending on how the worker service is scaled, this will still run multiple times.

It seems like DB backed job is the only sure way of doing this, other than using a redis backed distributed lock.

cc @eseliger @mrnugget

File: enterprise/cmd/frontend/internal/licensing/init/init.go (hunk 1 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/licensing/init/init.go:166]: Tbh its probably fine to move that to internal/, we're moving loads of stuff already so you could do a precursor PR to move it before this one

**Comment by Strum355** [comment]
Why are we moving this to worker btw? (i could search slack but itd be better for prosperity to have it mentioned here 😄)

**Comment by kopancek** [comment]
> Why are we moving this to worker btw? (i could search slack but itd be better for prosperity to have it mentioned here 😄)

The main reason was, that today if there are multiple instances of frontend, we fire concurrent requests to license check API.

There is some logic on the dotcom instance that relies on these requests to come at roughly 12 hour intervals. Receiving multiple requests causes this logic to break. There is other logic on the instance itself that relies on these calls to be done in 12 hour intervals. All of this can lead to concurrent DB/Redis writes, so there is a potential for race conditions, although I did not yet come up with a scenario where the concurrency itself would break the business logic.

All in all, doing the goroutine logic in a singleton would avoid the concurrency problem and seems safer. We can achieve it with a DB backed worker, but I wanted to try some quicker approaches first, because refactoring to DB backed worker would take more time and effort.

**Comment by mrnugget** [comment]
As far as I know, customers don't just duplicate `worker` instances but scale out specific workers by adding `worker` instance with worker names in `ALLOWLIST`. We have multiple jobs that don't make sense to be scaled out to more `worker` instances.

Just look through these: https://github.com/sourcegraph/sourcegraph/blob/d58eccf2cdfd792c979d2d57130535316669aff3/enterprise/cmd/worker/shared/shared.go#L34-L77

I think it's fine if you add a doc entry for your worker to https://docs.sourcegraph.com/admin/workers with the note that this should not be scaled out.

**Comment by eseliger** [comment]
Agree with @mrnugget, worker can either be scaled vertically (non-issue anyways), or horizontally, by splitting out jobs across multiple machines. That does by no means mean that three workers with the _same_ set of jobs should exist, it would be misconfiguration.

File: internal/licensing/worker.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+func (s *licenseWorker) Routines(_ context.Context, observationCtx *observation.Context) ([]goroutine.BackgroundRoutine, error) {
+	db, err := workerdb.InitDB(observationCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	gs, err := db.GlobalState().Get(context.Background())
+	if err != nil {
+		return nil, err
+	}
```
Comment (eseliger) [inline internal/licensing/worker.go:41]: I understand that this is never changing but should we disable the job entirely if the DB read fails once on startup?

Comment (eseliger) [reply internal/licensing/worker.go:41]: Also, if we keep this I think we can use the ctx that's passed down into `Routines`

File: internal/licensing/worker.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/env"
+	"github.com/sourcegraph/sourcegraph/internal/goroutine"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+
+	workerdb "github.com/sourcegraph/sourcegraph/cmd/worker/shared/init/db"
+)
+
+type licenseWorker struct{}
+
+func NewLicenseWorker() job.Job {
```
Comment (eseliger) [inline internal/licensing/worker.go:20]:
```suggestion
// NewLicenseWorker is the set of background jobs used for licensing enforcement and gating.
// Note: This job should only run once for a given Sourcegraph instance.
func NewLicenseWorker() job.Job {
```

**Review by eseliger — APPROVED** [review]

**Comment by mrnugget** [comment]
@kopancek what's the status here? We kinda forgot this in sprint planning

File: internal/licensing/worker.go (hunk 2 above)
Comment (kopancek) [reply internal/licensing/worker.go]: Moved the DB read to a different spot, so the worker will start even if DB is not initialized properly yet.

**Review by eseliger — APPROVED** [review]

**Pull request merged**